### ISSUE TITLE
Issuing parallel/asynchronous queries

### ISSUE THREAD

**Issue by jimenezrick** [twittner/cql-io#8]
Is it supposed to be possible to issue multiple queries using the same client with cql-io (0.13.2) from different threads? In my case if I try to to this I get the next error multiple times:

`2015-04-14T15:21:46Z, W, logger=cql, 127.0.0.1:9042, read-loop: threadWait: invalid argument (Bad file descrop)`

Also, is there any way to send queries asynchronously from the same thread?

As far as I know the CQL native protocol supports request pipelining so it would be a nice to have feature in this Haskell driver as it's present in other languages.

**Comment by twittner** [comment]
> Is it supposed to be possible to issue multiple queries using the same client with cql-io (0.13.2) from different threads? 

Yes.

> In my case if I try to to this I get the next error multiple times [...]

Could you please provide a bit more information regarding the setup used. Does the log contain more information on `Debug` or `Trace` level? Could you maybe even provide a test-case?

> Also, is there any way to send queries asynchronously from the same thread?

No.

> As far as I know the CQL native protocol supports request pipelining so it would be a nice to have feature in this Haskell driver as it's present in other languages.

Internally send and receive are concurrent operations. Subsequent sends do not have to wait for receives to complete.

**Comment by jimenezrick** [comment]
The thing is, I'm passing a list of requests and running the client in parallel in all of them with 
[mapConcurrently](https://hackage.haskell.org/package/async-2.0.2/docs/Control-Concurrent-Async.html#v:mapConcurrently) from `async` (actually I'm using `mapPool` from `async-extras` which is similar to mapConcurrently but with a bounded number of parallel tasks).

I tried using directly threads created manually by me and it works flawlessly. So, I need to look it further why it fails with `async` tasks. Given that I don't know the internals of that library, let me have a look to see the root cause.

**Comment by twittner** [comment]
Thanks. `mapConcurrently` should work. Could you maybe try the current `develop` branch which has some fixes w.r.t. connection and pool handling and check if it works for you. I tried the following:

```haskell
{-# LANGUAGE OverloadedStrings #-}

module Main (main) where

import Control.Concurrent.Async
import Data.Functor.Identity
import Data.Text (Text)
import Database.CQL.IO as Client
import Database.CQL.Protocol

import qualified System.Logger as Logger

main :: IO ()
main = do
    g <- Logger.new Logger.defSettings
    c <- Client.init g defSettings
    r <- mapConcurrently (runClient c) $ replicate 500 (query q p)
    mapM_ (print . map runIdentity) r

q :: QueryString R () (Identity Text)
q = "SELECT cql_version from system.local"

p :: QueryParams ()
p = QueryParams Quorum False () Nothing Nothing Nothing
```

**Comment by jimenezrick** [comment]
Great, thanks for the test case :), it reproduces precisely my issue, running that code with `cql-io` from `develop` always aborts the execution in the same way my application does:

```
$ ./test
2015-04-16T15:11:12Z, D, connecting to 127.0.0.1:9042
test: cql-io: hosts busy
```

Running all the queries from threads created manually works fine, so it fails whenever `async` is used. I was diving into the `async` code and as far as I can tell it's just a thin layer around threads creation and passing back the results.

The only thing I can think about is how `async` handles the masking of exceptions, could it be interfering with your library?

**Comment by twittner** [comment]
Can you double-check that you are on commit 8c782cb923b5828155f5be75e9c3f2dcdc8bd1cc? I noticed a problem with the pool implementation when trying out the test code from above and the fix thus fairly recent.

**Comment by jimenezrick** [comment]
I was testing with the previous revision, with 8c782cb923b5828155f5be75e9c3f2dcdc8bd1cc I cannot reproduce the issue any more, thanks!
